README.md
# **THIS BOT IS DEPRECATED**

[Zoom's September 21, 2020 release will have native breakout room switching without Host-assistance](https://www.reddit.com/r/Zoom/comments/irkm82/selfselect_breakout_room/). With that said, maybe the internals can be repurposed for something new that Zoom hasn't enabled? 

# Breakout Room Bot for Zoom Meetings Chrome Extension

*Making it possible for **everyone** to walk around to other Breakout Rooms in a Zoom Meeting without the aid of the Host*

Discord for help, comments, discussion, etc. : https://discord.gg/NS6nbrT

TODO: Video Showcase/Tutorial/Demo. Sample below; not final.

Example Screen Share in main meeting:
![](intro.png)
![](https://user-images.githubusercontent.com/5363/92406673-79d2e080-f0ed-11ea-9953-5b7704811d1c.gif)

## Usage (As Host)

It is strongly recommend you do a dry run with some associates first!

1. Install the extension from the Chrome Web Store @ https://chrome.google.com/webstore/detail/breakout-room-bot-for-zoo/acfkhlojnkihdmgikmkilfjkapkemcnd?hl=en&authuser=0
    * Alternatively, clone or download a copy of this repository and look for instructions online for loading "unpacked extensions" into Chrome or any other Chromium browser. Load the `extension` folder.
2. Go to the Zoom Meeting as the Host in a "Join from Browser" Zoom Meetings Web Client.
    * Button to launch extension only works on `*.zoom.us` domains.
3. Ensure the Breakout Rooms are already open and have started. The red Close All Rooms button should be present in the Breakout Rooms window.
4. Click the extension button in the menu bar to open the popup
    ![](launch.png)
5. Click the button to "Attach and Launch Breakout Room Bot"

After:

* Do not close the Chat Window while the bot is in operation.
* Unless you like cacophony, do not join audio in this client. You can leave audio by clicking the "^" next to the microphone icon and selecting "Leave `<Something>` Audio".
* Use another Zoom client on the system to actually participate in the meeting. You can safely make this other client a Co-Host if it isn't already from the attached bot-controlled web client UI.

### Notes for Co-Hosts

* With plain Zoom, you cannot assign users to breakout rooms. You can move around freely between Breakout Rooms without the bot though. Co-Hosts can also rename users. This means you can also rename users to have the bot assign them to Breakout Rooms!

## Usage (As Attendee)

**Chat commands only work in the main or root Zoom meeting**.

If you're in a breakout room, you can return to the main or root Zoom meeting from a Breakout room by selecting End Meeting in a Breakout Room and selecting Leave Breakout room.

* `!ls` List Rooms
    * Alternatively, the Host can share screen a slide or image with a list of room names.
    * This also takes an argument to list users in a particular room name.
* Switching Rooms as a non-host or non-cohost (this is whole reason for this bot's existence)
    * Two methods
        * *`!mv <room id>`* or *`!mv <room name>`*
            * Move to breakout room ID or a room name that best matches. List of room IDs and names of Breakout Rooms can be found with `!ls`. The host may have also screen shared a screen with the list.
        * *Rename yourself and append/change a `[room id]` or `[room name]` to your name.*
            * Unlike chat commands, this action *also works inside Breakout rooms and not just the main meeting*!
            * For example, if your name is "Bryan", change your name to "`Bryan [1]`" to move to the room with the ID of "`1`".
            * Another example, if your name is "Bryan", change your name to "`Bryan [Foo]`" to move to the room whose name best matches "Foo".
    * Both the `!mv` command and the name change action can take a *room name*.
      The best matching room name is the room the requester will be sent to. **This is a fuzzy search.**

      Breakout rooms are by default named `Breakout Room <some number>` but they can be renamed before the breakout rooms are opened or while the rooms are temporaily closed.
        * Example
            * If you have Breakout Rooms with Room IDs and named:
                1. Breakout Room 1
                2. Tokyo
                3. Los Angeles
                4. Las Vegas
                5. City 17
                6. Bangkok
                7. Pilgrims
                8. Breakout Room 15
                9. HNH1 computation group
                10. Henderson Party
            * To switch to "Las Vegas", "Bryan" can chat "`!mv las`" or change their name to "`Bryan [las]`".
                * Alternatively, the room ID of "Las Vegas" is 4 and Bryan can also chat
                  `!mv 4` or change their name to `Bryan [4]`.
        * If you don't get the room you want, just be more specific or "correct".
* `!version` Have the bot print out the version.

# Limits on Scale

* You should run this on a fast computer!
* If you have performance problem with native Zoom clients, close the chat window.

## Developer

* Zip the extension folder and upload it to the Chrome Web Store.
* It's a hack, apologies for the cobbled-togetherness.

## License

MIT

## Credit

This is likely cobbled together from piles of stackoverflow, random Google Searches, and contributors.

This hack/bot uses [RxJS][rxjs] and [Fuzzysort][fuzzysort].

[breakoutroominfo]: https://support.zoom.us/hc/en-us/articles/206476093-Enabling-breakout-rooms
[ocrbreakoutroombot]: https://github.com/ottoscholten/zoomChatBot
[desertpyhack]: https://www.meetup.com/Phoenix-Python-Meetup-Group/events/272227324/
[desertpy]: https://www.meetup.com/Phoenix-Python-Meetup-Group
[rxjs]: https://rxjs-dev.firebaseapp.com/
[fuzzysort]: https://github.com/farzher/fuzzysort
[preassign]: https://support.zoom.us/hc/en-us/articles/360032752671-Pre-assigning-participants-to-breakout-rooms


RELEASES.md
# 2020.09.7 2020-09-12

* Reworked bot to use internal Zoom command websocket connection to assign and reassign users to breakout rooms. ~100ms per assignment to 2ms per assignment! Greatly reduced safety latency to 10ms as a result.
* Buffered chat messages per second for assignment notices (26ms per second vs 26ms per user assignment)
* Enhanced `!ls` to also take arguments for room name to match and list users
* Enhanced `!ls` to list number of users in a breakout room

# 2020.09.6 2020-09-08

* Upon attachment, assign users by their names

# 2020.09.5 2020-09-07

* Lower safety latency for assignments to 150ms by removing some "sleeps".

# 2020.09.4 2020-09-07

* Tweak newlines in chat responses

# 2020.09.3 2020-09-07

* Change room by name search and id! Works in both renames and `!mv`.

# v1.4 2020-09-03

* Change Room by name changing. Only supports ID-indexes for now.

# v1.3 2020-09-03

* Permit all subdomains underneath zoom.us, allowing enterprise/school use and location-specific servers.


loadtest_extension/popup.js
// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

'use strict';

let launcher = document.getElementById('launcher');

launcher.onclick = function (element) {
    chrome.tabs.query(
        { active: true, currentWindow: true },
        function (tabs) {
            var url = chrome.runtime.getURL('LoadBot.js')
            setTimeout(_ => {
                chrome.tabs.executeScript(
                    tabs[0].id,
                    {
                        code: 'var bbSource = document.createElement("script");' +
                            `bbSource.setAttribute("src","${url}");` +
                            'document.head.appendChild(bbSource);'
                    }
                );
            }, 300);
        }
    );
    launcher.disabled = true;
}


loadtest_extension/LoadBot.js
function chatboxSend(msg) {
    const chatboxElement = document.getElementsByClassName('chat-box__chat-textarea')[0];
    const nativeTextAreaValueSetter = Object.getOwnPropertyDescriptor(window.HTMLTextAreaElement.prototype, "value").set;
    nativeTextAreaValueSetter.call(chatboxElement, msg);
    chatboxElement.dispatchEvent(new Event('input', { bubbles: true }));

    const oEvent = document.createEvent('KeyboardEvent');
    // Chromium Hack
    Object.defineProperty(oEvent, 'keyCode', {
        get: function () {
            return this.keyCodeVal;
        }
    });
    Object.defineProperty(oEvent, 'which', {
        get: function () {
            return this.keyCodeVal;
        }
    });

    const k = 13;

    oEvent.initKeyboardEvent("keydown", true, true, document.defaultView, k, k, "", "", false, "");

    oEvent.keyCodeVal = k;

    chatboxElement.dispatchEvent(oEvent);
}

async function renameSend(newName) {
    const newNameElement = await waitForElm("#newname")
    const nativeTextAreaValueSetter = Object.getOwnPropertyDescriptor(window.HTMLInputElement.prototype, "value").set;
    nativeTextAreaValueSetter.call(newNameElement, newName);
    newNameElement.dispatchEvent(new Event('input', { bubbles: true }));

    const oEvent = document.createEvent('KeyboardEvent');
    // Chromium Hack
    Object.defineProperty(oEvent, 'keyCode', {
        get: function () {
            return this.keyCodeVal;
        }
    });
    Object.defineProperty(oEvent, 'which', {
        get: function () {
            return this.keyCodeVal;
        }
    });

    const k = 13;

    oEvent.initKeyboardEvent("keydown", true, true, document.defaultView, k, k, "", "", false, "");

    oEvent.keyCodeVal = k;

    newNameElement.dispatchEvent(oEvent);
}

function reactMouseOver(el) {
    var oEvent = document.createEvent('MouseEvent');
    oEvent.initMouseEvent("mouseover", true, true, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null);
    el.dispatchEvent(oEvent);
}

// https://stackoverflow.com/a/61511955/286021
function waitForElm(selector) {
    return new Promise(resolve => {
        if (document.querySelector(selector)) {
            return resolve(document.querySelector(selector));
        }

        const observer = new MutationObserver(mutations => {
            if (document.querySelector(selector)) {
                resolve(document.querySelector(selector));
                observer.disconnect();
            }
        });

        observer.observe(document.body, {
            childList: true,
            subtree: true
        });
    });
}

function getRandomInt(min, max) {
    min = Math.ceil(min);
    max = Math.floor(max);
    return Math.floor(Math.random() * (max - min + 1)) + min;
}

// Open the chat pane if it isn't already open.
var chatPaneButton = document.querySelector('[aria-label^="open the chat pane"]')
if (chatPaneButton) {
    chatPaneButton.click();
}

// Open the participants pane if it isn't already open.
var participantsPaneButton = document.querySelector('[aria-label^="open the manage participants list pane"]')
if (participantsPaneButton) {
    participantsPaneButton.click();
}

reactMouseOver(document.querySelector('#participants-list-0'))

async function renameForTest() {
    var renameEl = await waitForElm('.button-margin-right.ax-outline-blue.btn.btn-xs.btn-primary')
    renameEl.click()
    await waitForElm('#newname')
    renameSend(`Load Test Bot #${getRandomInt(10000, 99999)}`)
}
renameForTest();

loadTestInteval = setInterval(() => {
    chatboxSend(`!mv ${getRandomInt(1, 10)}`)
}, 1000)

extension/BreakoutRoomBot.js
BREAKOUT_ROOM_BOT_VERSION = "2020.09.7"

// HELPER FUNCTIONS

function getStoreObservable(store) {
    return new rxjs.Observable(function (observer) {
        const unsubscribe = store.subscribe(function () {
            observer.next(store.getState());
        });
    });
}

function chatboxSend(msg) {
    const chatboxElement = document.getElementsByClassName('chat-box__chat-textarea')[0];
    const nativeTextAreaValueSetter = Object.getOwnPropertyDescriptor(window.HTMLTextAreaElement.prototype, "value").set;
    nativeTextAreaValueSetter.call(chatboxElement, msg);
    chatboxElement.dispatchEvent(new Event('input', { bubbles: true }));

    const oEvent = document.createEvent('KeyboardEvent');
    // Chromium Hack
    Object.defineProperty(oEvent, 'keyCode', {
        get: function () {
            return this.keyCodeVal;
        }
    });
    Object.defineProperty(oEvent, 'which', {
        get: function () {
            return this.keyCodeVal;
        }
    });

    const k = 13;

    oEvent.initKeyboardEvent("keydown", true, true, document.defaultView, k, k, "", "", false, "");

    oEvent.keyCodeVal = k;

    chatboxElement.dispatchEvent(oEvent);
}

function reactMouseOver(el) {
    var oEvent = document.createEvent('MouseEvent');
    oEvent.initMouseEvent("mouseover", true, true, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null);
    el.dispatchEvent(oEvent);
}

// This function uses websockets
function assignUserIdToBreakoutRoomUuid(senderUserId, roomUuid) {
    console.log(`Assigning ${senderUserId} to ${roomUuid}`)
    var storeState = internalStore.getState();
    var sender = storeState.attendeesList.attendeesList.find(({ userId }) => userId == senderUserId)
    var senderInBreakoutRoom = storeState.breakoutRoom.roomList.flatMap(room => room.attendeeIdList).includes(sender.userGUID)
    if (senderInBreakoutRoom) {
        window.commandSocket.send(JSON.stringify(
            { "evt": 4181, "body": { "targetBID": roomUuid, "targetID": senderUserId }, "seq": 0 }
        ))
    } else {
        window.commandSocket.send(JSON.stringify(
            { "evt": 4179, "body": { "targetBID": roomUuid, "targetID": senderUserId }, "seq": 0 }
        ))
    }
}

// https://stackoverflow.com/a/61511955/286021
function waitForElm(selector) {
    return new Promise(resolve => {
        if (document.querySelector(selector)) {
            return resolve(document.querySelector(selector));
        }

        const observer = new MutationObserver(mutations => {
            if (document.querySelector(selector)) {
                resolve(document.querySelector(selector));
                observer.disconnect();
            }
        });

        observer.observe(document.body, {
            childList: true,
            subtree: true
        });
    });
}

// OBSERVABLES

var internalStore = document.getElementById('root')._reactRootContainer._internalRoot.current.child.pendingProps.store;

var store$ = getStoreObservable(internalStore)

var chat$ = store$.pipe(
    rxjs.operators.map(s => s.chat.meetingChat),
    rxjs.operators.distinctUntilChanged(),
    rxjs.operators.skip(1),
)

// Transforms the user to map stuff to be a user -> name pair like IRC/etc
var userMessageMap$ = chat$.pipe(
    rxjs.operators.map(chatState => {
        return {
            sender: chatState.slice(-1)[0].sender,
            senderUserId: chatState.slice(-1)[0].senderId,
            message: chatState.slice(-1)[0].chatMsgs.slice(-1)[0],
        }
    }),
)

var versionCommand$ = userMessageMap$.pipe(
    rxjs.operators.filter(({ message }) => message == "!version"),
)

var versionReply$ = versionCommand$.pipe(
    rxjs.operators.map((_) => `🤖💔 BreakoutRoomBot ${BREAKOUT_ROOM_BOT_VERSION}\ngithub.com/nelsonjchen/HackyZoomBreakoutBot`)
)

var breakoutRoomListCommand$ = userMessageMap$.pipe(
    rxjs.operators.filter(({ message }) => message == "!ls"),
)

var breakoutRoomListReply$ = breakoutRoomListCommand$.pipe(
    rxjs.operators.withLatestFrom(
        store$,
        (_, storeState) =>
            "📜 Breakout Room List\n" +
            "Chat \"!ls\" to see this list\n" +
            "ID. Room Name (Attendee(s) in Room)\n=============\n" +
            storeState.breakoutRoom.roomList.map(
                (room, index) => `⬛ ${index + 1}. ${room.name} (${storeState.attendeesList.attendeesList.filter(attendee => attendee.bid == room.boId).length} attendee(s))`
            ).join('\n') +
            "\n" +
            "Chat \"!mv ID\" or append \"[ID]\" to your name to be assigned to the breakout room.\n" +
            "List attendees of a room by chatting \"!ls <room name here>\""
    ),
)

var breakoutRoomListUsersInRoomReply$ = userMessageMap$.pipe(
    rxjs.operators.filter(({ message }) => /!ls (.+)/.test(message)),
    rxjs.operators.withLatestFrom(
        store$,
        ({ message }, storeState) => {
            var targetRoomQuery = message.match(/!ls (.+)/)[1];

            var results = fuzzysort.go(targetRoomQuery, storeState.breakoutRoom.roomList, { key: 'name' });
            if (results.length == 0) {
                return `⚠️ No room names matched for query: ${targetRoomQuery}!\n`
            }
            var room = results[0].obj;
            var attendeesInRoom = storeState.attendeesList.attendeesList.filter(attendee => attendee.bid == room.boId)

            if (attendeesInRoom.length == 0) {
                return `📜 No attendees in ⬛ ${room.name}`
            }

            return `📜 Attendees in ⬛ ${room.name}\n` +
                attendeesInRoom.map(
                    attendee => `👤 ${attendee.displayName}`
                ).join('\n')
        }
    ),
)

var nameChange$ = store$.pipe(
    rxjs.operators.map(s => s.attendeesList.attendeesList),
    rxjs.operators.map(list => new Map(
        list.map(
            attendee => [attendee.userId, attendee.displayName]
        )
    )
    ),
    rxjs.operators.scan((acc, attendeesMap) => {
        if (acc === undefined) {
            return { previousMap: attendeesMap, changedNames: [] }
        }

        var changedNames = [];

        for (let [userId, displayName] of attendeesMap.entries()) {
            let oldDisplayName = acc.previousMap.get(userId);
            if (oldDisplayName != undefined && oldDisplayName != displayName) {
                changedNames.push({
                    oldDisplayName: oldDisplayName,
                    newDisplayName: displayName,
                    senderUserId: userId,
                });
            }
        }

        return { previousMap: attendeesMap, changedNames: changedNames }
    }, undefined),
    rxjs.operators.map((acc) => acc.changedNames),
    rxjs.operators.distinctUntilChanged(),
    rxjs.operators.filter((changedNames) => changedNames.length > 0),
    rxjs.operators.flatMap((changedNames) => rxjs.from(changedNames)),
)

var moveRequestQueryFromInitialNames$ = new rxjs.Subject();

var moveRequestQueryFromNameChange$ = nameChange$.pipe(
    // Only operate on names that change to a format requesting a room
    rxjs.operators.filter((changedNamePair) => {
        const regex = /\[(.+)\]/;
        return regex.test(changedNamePair.newDisplayName);
    }),
    // Ignore names that change but match the old name's query
    rxjs.operators.filter((changedNamePair) => {
        const regex = /\[(.+)\]/;
        var newNameTest = regex.test(changedNamePair.newDisplayName);
        var oldNameTest = regex.test(changedNamePair.oldDisplayName);
        if (newNameTest && oldNameTest) {
            var targetRoomNew = changedNamePair.newDisplayName.match(regex)[1]
            var targetRoomOld = changedNamePair.oldDisplayName.match(regex)[1]
            if (targetRoomNew == targetRoomOld) {
                return false;
            }
        }
        return true
    }),
    rxjs.operators.map((changedNamePair) => {
        const regex = /\[(.+)\]/;
        var targetRoomQuery = changedNamePair.newDisplayName.match(regex)[1]
        return {
            senderUserId: changedNamePair.senderUserId,
            sender: changedNamePair.newDisplayName,
            targetRoomQuery,
            src: 'nameChange'
        }
    }),
)

var chatMoveRequestCommand$ = userMessageMap$.pipe(
    rxjs.operators.filter(({ message }) => message.startsWith("!mv ")),
)

var moveRequestQueryFromChat$ = chatMoveRequestCommand$.pipe(
    rxjs.operators.map(({ sender, message, senderUserId }) => {
        const regex = /!mv (.+)/;
        var targetRoomQuery = message.match(regex)[1]
        return {
            sender: sender,
            senderUserId: senderUserId,
            targetRoomQuery: targetRoomQuery,
            src: 'chat'
        }
    }),
)

var moveRequestQuery$ = rxjs.merge(
    moveRequestQueryFromChat$,
    moveRequestQueryFromNameChange$,
    moveRequestQueryFromInitialNames$,
)

var [moveRequestSimpleIdQuery$, moveRequestStringQuery$] = moveRequestQuery$.pipe(
    rxjs.operators.partition(({ targetRoomQuery }) => /^\d+$/.test(targetRoomQuery))
)

var moveRequestSimpleIdQueryResolved$ = moveRequestSimpleIdQuery$.pipe(
    rxjs.operators.withLatestFrom(
        store$,
        ({ sender, targetRoomQuery, src, senderUserId }, storeState) => {
            var roomIndex = parseInt(targetRoomQuery, 10);

            if (roomIndex == 0 || roomIndex > storeState.breakoutRoom.roomList.length) {
                return { error: `⚠️ (from ${src})\n @${sender} Room ID "${targetRoomQuery}" out of range!\n` }
            }

            var room = storeState.breakoutRoom.roomList[roomIndex - 1]

            var roomName = room.name
            var roomUuid = room.boId

            return { sender, roomName, src, senderUserId, roomUuid }
        }
    ),
)

var moveRequestStringQueryResolved$ = moveRequestStringQuery$.pipe(
    rxjs.operators.withLatestFrom(
        store$,
        ({ sender, targetRoomQuery, src, senderUserId }, storeState) => {
            var results = fuzzysort.go(targetRoomQuery, storeState.breakoutRoom.roomList, { key: 'name' });
            if (results.length == 0) {
                return { error: `⚠️ (from ${src})\n @${sender} No names matched for query: ${targetRoomQuery}!\n` }
            }
            var roomName = results[0].obj.name;
            var roomUuid = results[0].obj.boId;

            return { sender, roomName, src, senderUserId, roomUuid }
        }
    ),
)

var moveRequestResolved$ = rxjs.merge(
    moveRequestSimpleIdQueryResolved$,
    moveRequestStringQueryResolved$,
)

var [moveRequestResolveError$, moveRequestResolved$] = moveRequestResolved$.pipe(
    rxjs.operators.partition(({ error }) => error),
)


var moveRequestChecked$ = moveRequestResolved$.pipe(
    rxjs.operators.withLatestFrom(
        store$,
        ({ sender, roomName, src, senderUserId, roomUuid }, storeState) => {
            var guidSenderMap = new Map(
                storeState.attendeesList.attendeesList.map(
                    attendee => [attendee.userGUID, attendee.displayName]
                )
            );

            var room = storeState.breakoutRoom.roomList.filter(room => room.name == roomName)[0]
            var roomAttendeesByName = room.attendeeIdList.map(attendeeId => guidSenderMap.get(attendeeId));

            if (roomAttendeesByName.includes(sender) && src == 'initialName') {
                // Don't return errors
                return null
            }

            if (roomAttendeesByName.includes(sender)) {
                return { error: `⚠️ (from ${src})\n "${sender}" already in "${room.name}"\n` }
            }

            return { sender, roomName, src, senderUserId, roomUuid }
        }
    ),
    rxjs.operators.filter(item =>
        item != null
    )
)

var [moveRequestInvalidError$, moveRequestValid$] = moveRequestChecked$.pipe(
    rxjs.operators.partition(({ error }) => error),
)

var moveRequestValidTimeSlice$ = rxjs.interval(10);

var moveRequestValidTimeSliceQueue$ = rxjs.zip(moveRequestValid$, moveRequestValidTimeSlice$).pipe(
    rxjs.operators.map(([s, _d]) => s)
)

var moveRequestError$ = rxjs.merge(
    moveRequestResolveError$,
    moveRequestInvalidError$,
)

var moveRequestErrorsAndSuccess$ = rxjs.merge(
    moveRequestError$,
    moveRequestValidTimeSliceQueue$
)

var moveFulfillChatResponse$ = new rxjs.Subject();

var moveFulfillChatResponseBuffered$ = moveFulfillChatResponse$.pipe(
    rxjs.operators.bufferTime(1000),
    rxjs.operators.filter(messages => messages.length != 0),
    rxjs.operators.map(messages => {
        if (messages.length == 1) {
            return messages[0]
        } else {
            return `🎯🎯🎯 Assigned ${messages.length} users over the last second.\n❓ Attendees may need to press the Breakout Rooms button to join their newly assigned breakout room.\n`
        }
    })
);

// SUBSCRIPTIONS

var versionReplySubscription = versionReply$.subscribe(
    (message) => chatboxSend(message)
)

var breakoutRoomListReplySubscription = breakoutRoomListReply$.subscribe(
    (message) => chatboxSend(message)
)

var breakoutRoomListReplySubscription = breakoutRoomListUsersInRoomReply$.subscribe(
    (message) => chatboxSend(message)
)

var moveRequestFulfillNotifySubscription = moveRequestErrorsAndSuccess$.subscribe(
    ({ sender, roomName, src, error, senderUserId, roomUuid }) => {
        if (error) {
            chatboxSend(error);
            return;
        }
        try {
            // ~ 2ms
            assignUserIdToBreakoutRoomUuid(senderUserId, roomUuid)
            moveFulfillChatResponse$.next(`🎯 (from ${src})\n Assigning\n👤 "${sender}"\nto\n⬛ "${roomName}"\n` +
                "❓ Attendee may need to press the Breakout Rooms button\n to join the newly assigned breakout meeting.\n❓ Chat \"!ls\" to list rooms and other commands.\n")
        } catch {

        }
    }
)

var moveFullfillChatResponseSubscription = moveFulfillChatResponseBuffered$.subscribe(message => {
    // ~ 30ms
    chatboxSend(message)
})

// Open the chat pane if it isn't already open.
var chatPaneButton = document.querySelector('[aria-label^="open the chat pane"]')
if (chatPaneButton) {
    chatPaneButton.click();
}

setTimeout(_ => {
    chatboxSend(`Breakout Rooms Bot for Zoom Meetings ${BREAKOUT_ROOM_BOT_VERSION} activated.\n\nAttendees, chat "!ls" in main meeting chat to list rooms and commands.\nUse a "!mv" in main meeting chat or append a [<room id>] or [room name] in to your name to choose a room.\nChat commands only work in main meeting\n however renames are detected anywhere.\nUse the End Meeting button in the Breakout Room to return to the main meeting in order to use chat commands\n\nHost(s), please rename and dedicate this client for the bot and use another session to participate in the meeting.`)
}, 100)

// Move users after bot initialization
setTimeout(_ => {
    internalStore.getState().attendeesList.attendeesList.map(
        attendee => attendee.displayName
    ).filter(displayName => /\[.+\]/.test(displayName)).map(
        name => {
            return {
                sender: name,
                targetRoomQuery: name.match(/\[(.+)\]/)[1],
                src: 'initialName'
            }
        }
    ).forEach(moveRequest => {
        moveRequestQueryFromInitialNames$.next(moveRequest);
    });
}, 110)
